marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/provenance/dh-provenance.xqy

```diff
@@ -0,0 +1,128 @@
+xquery version "1.0-ml";
+
+module namespace dhps = "http://marklogic.com/data-hub/dh-provenance-services";
+
+declare namespace prov = "http://www.w3.org/ns/prov#";
+declare namespace xs="http://www.w3.org/2001/XMLSchema";
+
+declare private variable $dhps-collection as xs:string := "http://marklogic.com/provenance-services/record";
+declare private variable $user-role as xs:string := "ps-user";
+declare private variable $internal-role as xs:string := "ps-internal";
+declare private variable $ps-dir as xs:string := "/provenance/";
+
+declare private function map-get($map as map:map, $key as xs:string){
+  let $items := map:get($map,$key)
+  return
+    if ($items instance of json:array) then
+      json:array-values($items)
+    else
+      $items
+};
+
+declare private function record-uri($provID as xs:string) as xs:string {
+  let $uri := $ps-dir||xdmp:sha256($provID)||".xml"
+  return $uri
+};
+
+declare private function wrap-in-prov-document($nodes as node()*) {
+  element prov:document {
+    namespace prov {"http://www.w3.org/ns/prov#"},
+    namespace xsi {"http://www.w3.org/2001/XMLSchema-instance"},
+    namespace ps  {"http://marklogic.com/provenance-services"},
+    namespace dh  {"http://marklogic.com/data-hub/prov#"},
+    namespace job  {"http://marklogic.com/data-hub/job#"},
+    namespace step  {"http://marklogic.com/data-hub/step#"},
+    $nodes
+  }
+};
+
+
+declare private function build-components($map as map:map) as node()*{
+  (
+    map:get($map,"activity"),
+    map:get($map,"agent"),
+    map:get($map,"wasAssociatedWith")
+  )
+};
+
+declare private function build($map as map:map) as node(){
+  wrap-in-prov-document(build-components($map))
+};
+
+declare private function job-record-activity($map as map:map, $id as xs:string, $options as map:map*) as map:map {
+  let $prov-id := fn:concat("job:", $id)
+  let $rec :=
+    element { xs:QName("prov:activity") } {
```

Comment:
Something to consider, but not necessarily change - I prefer to only use this form of XML construction when dynamically building elements and attributes - e.g. "element {fn:node-name($some-var)}". Otherwise, I think this is easier to read, with the downside being the need to duplicate QNames in start and close tags:

```
let $rec := 
  <prov:activity prov:id="{$prov-id}">
    <prov:label>{$id}<prov:label>
    etc...
```